# Clustering Neighbourhoods in Toronto
## Coursera Capstone Project 

### Parameswaran Periaswamy


### Part 1 - Converting Wikipedia html table into a DataFrame

In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' 

In [3]:
r=requests.get(my_url)

### Parsing the web html file with BeautifulSoup package

In [4]:
# Parse the htlm with Soup
page=bsoup(r.text,"html.parser")
page

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className = document.documentElement.className.replace( /(^|\s)client-nojs(\s|$)/, "$1client-js$2" );</script>
<script>(window.RLQ=window.RLQ||[]).push(function(){mw.config.set({"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":876823784,"wgRevisionId":876823784,"wgArticleId":539066,"wgIsArticle":true,"wgIsRedirect":false,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":false,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","

### 'The htlm 'page'  indicates that there is a 'table class' - let's find that table 

In [5]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
</td></tr>
<tr>
<td>M6A</td>

In [6]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

290

###  Such table (named 'results' contains 290 rows. The first tow is the table header.
### Therefore, there are 289 rows of data ( index 0 to 288)


In [7]:
results[0:5]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>]

### The Header of the DataFrame is the first row of data [index 0]

In [8]:
header=results[0].text.split()
header

['Postcode', 'Borough', 'Neighbourhood']

### Let's check some  rows in order to prepare to build the loop that will extract all cells into a DataFrame.  For example, let's examine row 85


In [9]:
results[85].text

'\nM5J\nDowntown Toronto\nToronto Islands\n'

In [10]:
results[85].text.split('\n')

['', 'M5J', 'Downtown Toronto', 'Toronto Islands', '']

In [11]:
results[85].text.split('\n')

['', 'M5J', 'Downtown Toronto', 'Toronto Islands', '']

In [12]:
Postcode=results[85].text.split('\n')[1]
Postcode

'M5J'

In [13]:
Borough=results[85].text.split('\n')[2]
Borough

'Downtown Toronto'

In [14]:
Neighborhood=results[85].text.split('\n')[3]
Neighborhood

'Toronto Islands'

## Iteration loop to extract all cells into a dataframe df

In [15]:
# iteration loop to harvest all records

records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)



PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

### Initial Dataframe has 289 cells or rows of data, as expected. Great!

In [16]:
df.shape

(289, 3)

### Let's check the tail of dataframe to assure all data is included as in Wikipedia table

In [17]:
df.tail()    #  Perfect! all data is included.

PostalCode       Borough          Neighbourhood
284        M8Z     Etobicoke              Mimico NW
285        M8Z     Etobicoke     The Queensway West
286        M8Z     Etobicoke  Royal York South West
287        M8Z     Etobicoke         South of Bloor
288        M9Z  Not assigned           Not assigned

## Now, let's remove all rows with Borough = 'Not assigned' strings
### There are 77 Boroughs with 'Not assigned' strings



In [18]:
# How many rows have Borough equal to 'Not assigned'?
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [19]:
# drops those rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [20]:
df1.head(5)

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

In [21]:
df1.tail()

PostalCode    Borough             Neighbourhood
207        M8Z  Etobicoke  Kingsway Park South West
208        M8Z  Etobicoke                 Mimico NW
209        M8Z  Etobicoke        The Queensway West
210        M8Z  Etobicoke     Royal York South West
211        M8Z  Etobicoke            South of Bloor

### The dataframe now has 212 rows (0-211) after droppong all Borough with 'Not assigned', which is correct ( 289-77) 

In [22]:
df1.shape

(212, 3)

### Now,  let's replace  'Not assigned' neighborhoods with the name of the Borough


In [23]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [24]:
df1.head()           # Now, there are no rows with 'Not assigned' strings anymore

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

In [25]:
df1.shape

(212, 3)

## Now , we will consolidate the dataframe to the unique Postal Codes.
### As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged.


### Let's examine de dataframe first to set up the best loop algorithm. How many unique postalcodes , Boroughs and Neighbourhood names are there?



In [26]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

     

Unique Postalcodes : 103
Unique Boroughs  : 11
Unique Neighbourhoods  :210


### There are 103 Unique Postcodes and 11 Unique Borough with 210 Neighbourhoods

In [27]:
#df1.groupby(['PostalCode','Borough','Neighbourhood']).size().reset_index(name='Count').head()

## Let's Consolidate the dataframe to each unique PostalCodes and aggregated Neighbourhoods

In [28]:
# Starting DataFrame df1
df1.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

In [29]:
nrows1=len(df1)
nrows1

212

### Let's keep df1 intact and let's work with df2 as precaution

In [30]:
# DataFrame df2 also has 212 rows or cells
df2=df1
df2.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

In [31]:
# I will use two pointers, 'n and m' to sweep through the dataframe. So 'n' will vary from
#  0 to 211, and 'm ' will vary from 1 to 212
nrows2=len(df2)-1
nrows2

211

## Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes.
### We expect a df2 with 103 rows because that is the number of unique postalcodes

In [32]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index


RangeIndex(start=0, stop=103, step=1)

In [33]:
df2.head()

PostalCode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   

                                       Neighbourhood  
0                                          Parkwoods  
1                                   Victoria Village  
2  Kingsway Park South West,Mimico NW,The Queensw...  
3                    Lawrence Heights,Lawrence Manor  
4                                       Queen's Park

In [34]:
df2.index

RangeIndex(start=0, stop=103, step=1)

## DataFrame consolidated to unique PostalCodes 
##  103 rows (0 to 102)  by 3 columns

In [35]:
df2.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2   Kingsway Park South West,Mimico NW,The Queensw...  
3                     Lawrence Heights,Lawrence Manor  
4                                        Queen's Park  
5                                    Islington Avenue  
6                                       Rouge,Malvern  
7                                     Don Mills North  
8                      Woodbine Gardens,Parkview Hill  
9                             Ryerson,Garden District  
10                                          Glencairn  
11  Cloverdale,Islington,Martin Grove,Princess Gar...  
12               Highland Creek,Rouge Hill,Port Union  
13                    Flemingdon Park,Don Mills South  
14                                   Woodbine Heights

In [36]:
df2.tail()

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
98         The Kingsway,Montgomery Road,Old Mill North  
99                                Church and Wellesley  
100  Business Reply Mail Processing Centre 969 Eastern  
101  Humber Bay,King's Mill Park,Kingsway Park Sout...  
102  Kingsway Park South West,Mimico NW,The Queensw...

### Dataframe has 103 rows and 3 columns

In [37]:
df2.shape

(103, 3)

### End of Part 1

### x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x

## PART 2: 
### Now, let's get the Latitude and Longitude coordenates for every PostalCode and add those columns to the dataframe

## NOTE:     I tried the geocoder but it did not work reliably, therefore I decided to use the cvs file with postalcodes provided.

In [38]:
# Read the cvs file and convert it to a dataframe
#df_pcodes=pd.read_csv('Toronto_Geospatial_Coordinates.csv')
#df_pcodes.columns = ['PostalCode','Latitude','Longitude']
#df_pcodes.head()

# Read the cvs file and convert it to a dataframe

url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Need to rename the column "Postal Code" to "PostalCode" in order to do a proper merger

In [39]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Now, let's sort our db2 dataframe by PostalCode, then we merge it with the df_pcodes created from the cvs file, based on common column of PostalCodes



In [40]:
df2s=df2.sort_values('PostalCode')

## Table  - 103 rows by 5 columns - as required
### dataframe 'neighborhoods' contains the latitude and longitude needed to start mapping.

### Table is sorted by PostalCode

In [41]:
neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'PostalCode')
neighborhoods.head(15)

PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                      Rouge,Malvern   
1         M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2         M1E  Scarborough                    Guildwood,Morningside,West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park   
7         M1L  Scarborough                      Clairlea,Golden Mile,Oakridge   
8         M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West   
9         M1N  Scarborough                         Birch Cliff,Cliffside West   
10        M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...   
11        M1R  Scarborough                                   Maryvale,Wexford   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter   
14        M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

In [42]:
neighborhoods.shape

(103, 5)

### END OF PART 2

### x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x

## PART 3:

## Explore and cluster the neighborhoods in Toronto

In [43]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


## Note
### I will focus on a subset of all the Borough that contain the word "Toronto". 
### There are 38 Borough with word Toronto.

In [44]:
toronto_data= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_data.head()

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [45]:
toronto_data.shape

(38, 5)

### From google :  the latitud and longitud for Toronto, as follows: 43.6532° N, 79.3832° W

In [46]:
latitude = 43.6532
longitude= -79.3832

### Let's create a map of Toronto with Boroughs that cointain the word Toronto

In [47]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.
Define Foursquare Credentials and Version

### Let's create a function to repeat the same process to all the neighborhoods in Manhattan Lab example

In [49]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [58]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            'ENK1ML0LUPURGPBWNIG5BKBTICO42RZWDOXFLT2AIB2AF14I', 
            'Q4JQG5FJZBJWC1JS35NVO524KVUCK24320VAQFBC5HVNQQSQ', 
            '20180605', 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [59]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Kingsway Park South West,Mimico NW,The Queensway West,Royal York South West,South of Bloor
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place

In [60]:
print(toronto_venues.shape)
toronto_venues.head()

(1687, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0  The Big Carrot Natural Food Market       43.678879       -79.297734   
1                 Grover Pub and Grub       43.679181       -79.297215   
2                           Starbucks       43.678798       -79.298045   
3                 Glen Stewart Ravine       43.676300       -79.294784   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0     Health Food Store  
1                   Pub  
2           Coffee Shop  
3  Other Great Outdoors  
4          Neighborhood

### Let's check how many venues were returned for each neighborhood

In [61]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            58   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     13   
Cabbagetown,St. James Town                                             46   
Central Bay Street                                                     81   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               15   
Church and Wellesley                                                   84   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             33   
Davisville North                                                        7   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            20   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      5   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           25   
Kingsway Park South West,Mimico NW,The Queenswa...                     46   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                62   
Moore Park,Summerhill East                                              1   
North Toronto West                                                     18   
Parkdale,Roncesvalles                                                  16   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede,Swansea                                                      35   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        40   
The Annex,North Midtown,Yorkville                                      24   
The Beaches                                                             5   
The Beaches West,India Bazaar                                          21   
The Danforth West,Riverdale                                            43   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             58   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      13   
Cabbagetown,St. James Town                                              46   
Central Bay Street                                                      81   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    84   
Commerce Court,Victoria Hotel      

### Let's find out how many unique categories can be curated from all the returned venues

In [62]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


## Analyze Each Neighborhood

In [63]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0   The Beaches                  0                  0        0   
1   The Beaches                  0                  0        0   
2   The Beaches                  0                  0        0   
3   The Beaches                  0                  0        0   
4   The Beaches                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0           0                    0                 0              0   
1           0                    0                 0              0   
2           0                    0                 0              0   
3           0                    0                 0              0   
4           0                    0                 0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0        

In [64]:
toronto_onehot.shape

(1687, 235)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [65]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                              Adelaide,King,Richmond               0.01   
1                                         Berczy Park               0.00   
2          Brockton,Exhibition Place,Parkdale Village               0.00   
3   Business Reply Mail Processing Centre 969 Eastern               0.00   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...               0.00   
5                          Cabbagetown,St. James Town               0.00   
6                                  Central Bay Street               0.00   
7             Chinatown,Grange Park,Kensington Market               0.00   
8                                            Christie               0.00   
9                                Church and Wellesley               0.00   
10                      Commerce Court,Victoria Hotel               0.00   
11                                         Davisville               0.00   
12                                   Davisville North               0.00   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...               0.00   
14            Design Exchange,Toronto Dominion Centre               0.00   
15                        Dovercourt Village,Dufferin               0.00   
16              First Canadian Place,Underground city               0.00   
17                 Forest Hill North,Forest Hill West               0.00   
18                      Harbord,University of Toronto               0.00   
19    Harbourfront East,Toronto Islands,Union Station               0.00   
20                       High Park,The Junction South               0.00   
21  Kingsway Park South West,Mimico NW,The Queensw...               0.00   
22                                      Lawrence Park               0.00   
23                            Little Portugal,Trinity               0.00   
24                         Moore Park,Summerhill East               0.00   
25                                 North Toronto West               0.00   
26                              Parkdale,Roncesvalles               0.00   
27                                           Rosedale               0.00   
28                                           Roselawn               0.00   
29                                  Runnymede,Swansea               0.00   
30                            Ryerson,Garden District               0.00   
31                                     St. James Town               0.01   
32                    Stn A PO Boxes 25 The Esplanade               0.00   
33                                    Studio District               0.00   
34                  The Annex,North Midtown,Yorkville               0.00   
35                                        The Beaches               0.00   
36                      The Beaches West,India Bazaar               0.00   
37                        The Danforth West,Riverdale               0.00   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.076923            0.076923      0.076923   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011905  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000         

#### Let's confirm the new size

In [66]:
toronto_grouped.shape

(38, 235)

### Let's print each neighborhood along with the top 5 most common venues

In [67]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.06
1             Café  0.04
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2      Farmers Market  0.03
3          Restaurant  0.03
4  Seafood Restaurant  0.03


----Brockton,Exhibition Place,Parkdale Village----
               venue  freq
0     Breakfast Spot  0.09
1               Café  0.09
2        Coffee Shop  0.09
3  Convenience Store  0.05
4      Grocery Store  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1                 Spa  0.06
2       Auto Workshop  0.06
3                Park  0.06
4          Comic Shop  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.15
1   Airport Service  0.

### Let's put that into a pandas dataframe

#### First, let's write a function to sort the venues in descending order.

In [68]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [69]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
5                          Cabbagetown,St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7             Chinatown,Grange Park,Kensington Market                  Café   
8                                            Christie         Grocery Store   
9                                Church and Wellesley           Coffee Shop   
10                      Commerce Court,Victoria Hotel           Coffee Shop   
11                                         Davisville          Dessert Shop   
12                                   Davisville North                 Hotel   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
14            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                        Dovercourt Village,Dufferin              Pharmacy   
16              First Canadian Place,Underground city           Coffee Shop   
17                 Forest Hill North,Forest Hill West                 Trail   
18                      Harbord,University of Toronto                  Café   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
20                       High Park,The Junction South         Grocery Store   
21  Kingsway Park South West,Mimico NW,The Queensw...           Coffee Shop   
22                                      Lawrence Park                  Park   
23                            Little Portugal,Trinity                   Bar   
24                         Moore Park,Summerhill East            Playground   
25                                 North Toronto West   Sporting Goods Shop   
26                              Parkdale,Roncesvalles        Breakfast Spot   
27                                           Rosedale                  Park   
28                                           Roselawn                Garden   
29                                  Runnymede,Swansea           Coffee Shop   
30                            Ryerson,Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                  The Annex,North Midtown,Yorkville                  Café   
35                                        The Beaches     Health Food Store   
36                      The Beaches West,India Bazaar        Sandwich Place   
37                        The Danforth West,Riverdale      Greek Restaurant   

   2nd Most Common Venue  3rd Most Common Venue  \
0             Steakhouse                    Bar   
1           Cocktail Bar         Farmers Market   
2         Breakfast Spot                   Café   
3         Farmers Market          Auto Workshop   
4        Airport Service       Airport Terminal   
5             Restaurant     Italian Restaurant   
6     Italian Restaurant                   Café   
7                    Bar  Vietnamese Restaurant   
8                   Café                   Park   
9    Japanese Restaurant       Sushi Restaurant   
10                  Café                  Hotel   
11        Sandwich Place            Coffee Shop   
12        Breakfast Spot           Burger Joint   
13                   Pub       Sushi Restaurant   
14                  Café                  Hotel   
15                Bakery         Discount Store   
16                  Café                  Hotel   
17                  Park       Sushi Res

## Cluster Neighbourhoods

#### Run k-means to cluster the neighbourhood into 5 clusters.

In [70]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [71]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store  Other Great Outdoors   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0        Sandwich Place             Pet Store   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               0                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop                   Pub                 Diner   
1        Ice Cream Shop    Italian Restaurant         Grocery Store   
2                  Park            Board Shop               Brewery   
3    Italian Restaurant                Bakery   American Restaurant   
4              Bus Line           Yoga Studio        Discount Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Farmers Market    Falafel Restaurant           Event Space   
1             Bookstore               Brewery       Bubble Tea Shop   
2          Burger Joint         Burrito Place                   Pub   
3             Gastropub           Cheese Shop           Fish Market   
4        Farmers Market    Falafel Restaurant           Event Space   

  9th Most Common Venue 10th Most Common Venue  
0  Ethiopian Restaurant      Electronics Store  
1                  Café             Restaurant  
2           Pizza Place          Movie Theater  
3             Juice Bar              Bookstore  
4  Ethiopian Restaurant      Electronics Store

# Final  resulting clusters visualization - Toronto Neighborhoods

In [72]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Extra item - Example of Examining a cluster:

#### Now, one can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, one can then assign a name to each cluster, if so desired

### Cluster 1

### Coffee shops, restaurant , pizza places and bookstores are common in this cluster

In [73]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0     Health Food Store   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0        Sandwich Place   
3       East Toronto               0                  Café   
4    Central Toronto               0                  Park   
5    Central Toronto               0                 Hotel   
6    Central Toronto               0   Sporting Goods Shop   
7    Central Toronto               0          Dessert Shop   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
23   Central Toronto               0                 Trail   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0           Coffee Shop   
34      West Toronto               0         Grocery Store   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0           Coffee Shop   
37      East Toronto               0    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  \
0   Other Great Outdoors            Coffee Shop   
1            Coffee Shop         Ice Cream Shop   
2              Pet Store                   Park   
3            Coffee Shop     Italian Restaurant   
4            Swim School               Bus Line   
5         Breakfast Spot           Burger Joint   
6            Coffee Shop         Clothing Store   
7         Sandwich Place            Coffee Shop   
8                  Diner   Fast Food Restaurant   
9                    Pub       Sushi Restaurant   
10            Playground                  Trail   
11            Restaurant     Italian Restaurant   
12   Japanese Restaurant       Sushi Restaurant   
13                  Café                 Bakery   
14        Clothing Store                   Café   
15            Restaurant                   Café   
16          Cocktail Bar         Farmers Market   
18            Steakhouse                    Bar   
19              Aquarium                  Hotel   
20                  Café                  Hotel   
21                  Café                  Hotel   
23                  Park       Sushi Restaurant   
25                   Bar    Japanese Restaurant   
26                   Bar  Vietnamese Restaurant   
29                  Café                  Hotel   
30                  Café                   Park   
31                Bakery         Discount Store   
32           Coffee Shop       Asian Restaurant   
33        Breakfast Spot                   Café   
34                   Bar                   Café   
35             Gift Shop     Italian Restaurant   
36           Pizza Place                   Café   
37        Farmers Market          Auto Workshop   

            4th Most Common Venue      5th Most Common Venue  \
0                             Pub                      Diner   
1              Italian Restaurant        

## End of part 3